blocks: support saving of all tags on tag debug

igorauad opened the pull request:
For testing, it is often useful to check all tags ever received, rather than
solely the tags from the last call to the Tag Debug's work function. This patch
adds this option and a public method for activating it.

This is more appropriate for controlled test environments, where the vector of
tags is known not to grown indefinitely. Hence, the option is not exposed as a
block parameter.

mbr0wn commented on gr-blocks/lib/tag_debug_impl.cc:
```diff
@@ -57,7 +58,11 @@ tag_debug_impl::~tag_debug_impl() {}
     gr::thread::scoped_lock l(d_mutex);
-    return d_tags;
+
+    if (d_save_all)
```
Please don't skimp on the {}.

igorauad replied:
Updated

mbr0wn commented on gr-blocks/lib/tag_debug_impl.cc:
```diff
@@ -69,6 +74,13 @@ int tag_debug_impl::num_tags()
 void tag_debug_impl::set_display(bool d) { d_display = d; }
 
+void tag_debug_impl::set_save_all(bool s) {
+    if (d_save_all && !s)
```
```suggestion
    if (d_save_all && !s) {
```

igorauad replied:
Updated

igorauad commented:
> Very useful. Please make sure your code matches our clang-format settings, I think it might not. Also, we need your CLA :) Thanks for the submission!

OK @mbr0wn I've ran `clang-format` and pushed the updates.

Regarding the CLA, I will ping you once it is arranged.

igorauad commented:
@mbr0wn I have finally got the CLA assigned, which I just sent to Ben over email. So I will rebase my pending MRs and remove the CLA-needed label from them.

mbr0wn commented on gr-blocks/lib/tag_debug_impl.cc:
```diff
@@ -57,6 +63,15 @@ int tag_debug_impl::num_tags()
+void tag_debug_impl::set_save_all(bool s)
+{
+    if (d_save_all && !s) {
+        d_tags.clear();
```
One more thing: Can you hold `d_setlock` while you access `d_tags`. Otherwise, this is not thread-safe.

igorauad replied:
Thanks, @mbr0wn . Tag Debug already had mutex `d_mutex`, which is locked both in `work()` and `current_tags()`. It was missing on `set_save_all()` and I've just added it.

mbr0wn commented on gr-blocks/lib/tag_debug_impl.cc:
```diff
@@ -109,6 +123,11 @@ int tag_debug_impl::work(int noutput_items,
+
+        if (d_save_all) {
+            for (const auto& tag : d_work_tags)
+                d_tags.push_back(tag);
```
Same here.

igorauad replied:
I believe the `gr::thread::scoped_lock l(d_mutex);` in the beginning of the `work` function suffices.

mbr0wn commented:
@igorauad Thanks for the update. Other than the thread-safety, this is good and I think it's a great addition to the code.

igorauad commented:
> @igorauad Thanks for the update. Other than the thread-safety, this is good and I think it's a great addition to the code.

Thanks, @mbr0wn . Updated and rebased.

michaelld commented:
@mbr0wn please give this one last read-through & if you approve then please merge
